File: 1. 단계별로_풀어보기/20~29/24. 분할 정복/7. 10830.py
import sys

input = sys.stdin.readline
output = sys.stdout.write

N, B = map(int, input().split())
A = list(list(map(int, input().split())) for i in range(N))

def matMul(matA, matB):
  maxN = len(matA)
  maxM = len(matB)
  maxK = len(matB[0])

  result = list(list(0 for k in range(maxK)) for n in range(maxN))

  for n in range(maxN):
    for k in range(maxK):
      for m in range(maxM):
        result[n][k] += matA[n][m] * matB[m][k]
        result[n][k] %= 1000

  return result

expArr = [0] * 48
expArr[0] = A

for i in range(1, 48):
  expArr[i] = matMul(expArr[i - 1], expArr[i - 1])

result = list(list(0 for i in range(N)) for j in range(N))

for i in range(N):
  result[i][i] = 1

idx = 0

while B > 0:
  if B & 1 == 1:
    result = matMul(result, expArr[idx])
  
  B //= 2
  idx += 1

for row in result:
  output(f"{' '.join(str(num) for num in row)}\n")

File: 1. 단계별로_풀어보기/20~29/24. 분할 정복/7. 10830.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const [N, B] = inputArr[0].trim().split(' ').map(BigInt)
const A = inputArr.slice(1).map(row => row.trim().split(' ').map(Number))

// 거듭제곱 저장할 배열
const expArr = Array.from({length: 48}, () => [])

// 배열 초기값 설정
expArr[0] = A

// 배열값 연산
for (let i = 1; i < 48; i++) {
  expArr[i] = matMul(expArr[i - 1], expArr[i - 1])
}

let idx = 0
let remain = B
let result = Array.from({length: +N.toString()}, () => Array.from({length: +N.toString()}).fill(0))

// 단위 행렬로 만들기
result.forEach((row, idx) => row[idx] = 1)

while (remain > 0n) {
  if (remain & 1n) {
    result = matMul(result, expArr[idx])
  }
  idx += 1
  remain = remain >> 1n
}

// 출력
result.forEach(row => {
  console.log(`${row.join(' ')}`)
})


function matMul(matA, matB) {
  const maxN = matA.length
  const maxM = matB.length
  const maxK = matB[0].length

  const result = Array.from({length: maxN}, () => Array.from({length: maxK}).fill(0))

  for (let n = 0; n < maxN; n++) {
    for (let k = 0; k < maxK; k++) {
      for (let m = 0; m < maxM; m++) {
        result[n][k] += matA[n][m] * matB[m][k]
        result[n][k] %= 1000
      }
    }
  }

  return result
}

